### PR TITLE
Add the function of generating correlated graph pairs and test.

### LINKED ISSUE #66: sample correlated SBM
Example:

require(igraph)
gg <- rg.sample.SBM.correlated(n = 100, B = matrix(c(0.5,0.5,0.2,0.5), nrow = 2), rho = c(0.4,0.6), sigma = 0.2)

summary(gg$adjacency$A)
IGRAPH c77bf6c U--- 100 2424 -- 
summary(gg$adjacency$B)
IGRAPH 3ccfb0c U--- 100 2039 -- 

cor(as.vector(gg$adjacency$A[]), as.vector(gg$adjacency$B[]))
[1] 0.1494246

## Correlated ER
rg.sample.correlated.gnp <- function(P,sigma){
   require(igraph)
   n <-  nrow(P)
   U <- matrix(0, nrow = n, ncol = n)
   U[col(U) > row(U)] <- runif(n*(n-1)/2)
   U <- (U + t(U))
   diag(U) <- runif(n)
   A <- (U < P) + 0 ;
   diag(A) <- 0

   avec <- A[col(A) > row(A)]
   pvec <- P[col(P) > row(P)]
   bvec <- numeric(n*(n-1)/2)

   uvec <- runif(n*(n-1)/2)

   idx1 <- which(avec == 1)
   idx0 <- which(avec == 0)

   bvec[idx1] <- (uvec[idx1] < (sigma + (1 - sigma)*pvec[idx1])) + 0
   bvec[idx0] <- (uvec[idx0] < (1 - sigma)*pvec[idx0]) + 0

   B <- matrix(0, nrow = n, ncol = n)
   B[col(B) > row(B)] <- bvec
   B <- B + t(B)
   diag(B) <- 0

   return(list(A = graph.adjacency(A,"undirected"), B = graph.adjacency(B,"undirected")))
}

# non-igraph version of correlated SBM
#gg <- rg.sample.SBM.correlated(n = 100, B = matrix(c(0.5,0.5,0.2,0.5), nrow = 2), rho = c(0.4,0.6), sigma = 0.2)
#cor(as.vector(gg$adjacency$A[]), as.vector(gg$adjacency$B[]))
rg.sample.SBM.correlated <- function(n, B, rho, sigma, conditional = FALSE){
   if(!conditional){
       tau <- sample(c(1:length(rho)), n, replace = TRUE, prob = rho)
   }
   else{
       tau <- unlist(lapply(1:2,function(k) rep(k, rho[k]*n)))
   }
   P <- B[tau,tau]
   return(list(adjacency=rg.sample.correlated.gnp(P, sigma),tau=tau))
}

### PR DESCRIPTION
**Pull request by SHAAAAN**
Closes #66 

This Pull Request includes functions, tests, tutorial, docstrings of generating correlated graph pairs with the same probability of observing the same edge in each graph pair with ER and SBM models. 

There is also a notebook to show the properties of the functions. The probability can be ranged from 0.01 to 0.99 and the correlation can be ranged from -0.99 to 0.99. 
https://nbviewer.jupyter.org/github/NeuroDataDesign/team-networkclass/blob/master/Shan/notebook_corr.ipynb
Here is another proof notebook for corr_sbm function:
https://nbviewer.jupyter.org/github/SHAAAAN/MyGraspy/blob/master/CorrectedCorrSBM.ipynb

### CODE REVIEW

**Comment by SHAAAAN** [comment]
@bdpedigo Could you please review this PR and leave some comments for me?

File: graspy/simulations/simulations_corr.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+def sample_edges_corr(P, Rho, directed, loops):
+    """
+    Generate a pair of correlated graphs with Bernoulli distribution.
+    Both G1 and G2 are binary matrices. 
+
+    Parameters
+    ----------
+    P: np.ndarray, shape (n_vertices, n_vertices)
+        Matrix of probabilities (between 0 and 1) for a random graph
+    Rho: np.ndarray, shape (n_vertices, n_vertices)
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: Can you call this `R` or something like that? Typically in python variables should be [snake case](https://en.wikipedia.org/wiki/Snake_case) but often this is neglected in the case of matrices (which are sometimes just capital letters) but otherwise I wouldn't have a variable start with capital letter unless it is a class

File: graspy/simulations/simulations_corr.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    directed: boolean, optional (default=False)
+        If False, output adjacency matrix will be symmetric. Otherwise, output adjacency
+        matrix will be asymmetric.
+    loops: boolean, optional (default=False)
+        If False, no edges will be sampled in the diagonal. Otherwise, edges
+        are sampled in the diagonal.
+
+    References
+    ----------
+    .. [1] Vince Lyzinski, et al. "Seeded Graph Matching for Correlated Erd}os-Renyi Graphs",
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: remove the } in erdos unless it does something

File: graspy/simulations/simulations_corr.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    References
+    ----------
+    .. [1] Vince Lyzinski, et al. "Seeded Graph Matching for Correlated Erd}os-Renyi Graphs", 
+       Journal of Machine Learning Research 15, 2014
+        
+    Returns
+    -------
+    G1: ndarray (n_vertices, n_vertices)
+        Binary matrix the same size as P representing a random graph,
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: adjacency matrix

File: graspy/simulations/simulations_corr.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        generated by the function of sample_edges
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: don't need to say what function generated it

File: graspy/simulations/simulations_corr.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    G2: ndarray (n_vertices, n_vertices)
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: same 2 comments as for G1

File: graspy/simulations/simulations_corr.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            [0., 0., 0., 0., 1.],
+            [0., 0., 1., 1., 0.]]), array([[0., 1., 0., 0., 0.],
+            [1., 0., 1., 0., 1.],
+            [0., 1., 0., 1., 1.],
+            [0., 0., 1., 0., 1.],
+            [0., 1., 1., 1., 0.]]))
+    """
+    G1 = sample_edges(P, directed=directed, loops=loops)
+    G0 = G1.copy()
+    G0 = np.where(G0 == 1, P + Rho * (1 - P), P * (1 - Rho))
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: can you change `G0` to `P2` or something like that? It is not a graph

File: tests/test_sims_corr.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        g1, g2 = sample_edges_corr(P, Rho, directed=False, loops=False)
+        # check the 1 probability of the output binary matrix
+        # should be near to the input P matrix
+        self.assertTrue(
+            np.isclose(self.p, g2.sum() / (self.n * (self.n - 1)), atol=0.05)
+        )
+        # check rho in graph2
+        add = g1 + g2
+        add[add != 2] = 0
+        real_prob = add.sum() / (2 * self.n * (self.n - 1))
```
Comment (bdpedigo) [inline tests/test_sims_corr.py]: change `real_` to expected

File: graspy/simulations/simulations_corr.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    Parameters
+    ----------
+    n: int
+       Number of vertices
+
+    p: float
+        Probability of an edge existing between two vertices, between 0 and 1.
+    
+    R: np.ndarray, shape (n_vertices, n_vertices)
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: can you make R a scalar for this function, apologies if that was not clear. Just like p vs P, for this function make r a single scalar

File: graspy/simulations/simulations_corr.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+def sample_edges_corr(P, R, directed, loops):
+    """
+    Generate a pair of correlated graphs with Bernoulli distribution.
+    Both G1 and G2 are binary matrices. 
+
+    Parameters
+    ----------
+    P: np.ndarray, shape (n_vertices, n_vertices)
+        Matrix of probabilities (between 0 and 1) for a random graph
+    R: np.ndarray, shape (n_vertices, n_vertices)
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: you need to do input checks, mainly just for R. What possible values of correlation are allowed? What if someone inputs a string instead of a matrix?

File: tests/test_sims_corr.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
+import unittest
```
Comment (bdpedigo) [inline tests/test_sims_corr.py]: you will also need tests once you write input checks

Comment (bdpedigo) [reply tests/test_sims_corr.py]: tests that make sure that wrong values of the parameters are not allowed by the function. See examples in the simulations tests of how to do this

File: graspy/simulations/simulations_corr.py (hunk 8 above)
Comment (bdpedigo) [reply graspy/simulations/simulations_corr.py]: @SHAAAAN keep in mind this was just for the `er_corr` function

**Comment by SHAAAAN** [comment]
Here is a notebook showing the properties of sample_edges_corr, er_corr and sbm_corr functions.
https://nbviewer.jupyter.org/github/NeuroDataDesign/team-networkclass/blob/master/Shan/notebook_corr.ipynb

**Comment by bdpedigo** [comment]
Hi @SHAAAAN, nice progress. I still need to take a closer look but a few things at first glance
 - all of the most recent commits to graspy are shown as new commits here, how did you go about keeping things up to date with master? Did you do a rebase or merge? We don't want those changes reflected as commits that you made as part of this PR (since they are already in graspy) so we need to revert those changes in your branch, ideally
 - hard for me to tell if the SBM is working, graph 1 and graph 2 look a bit off. In terms of probability and maybe rho. have you tested against Bear's code at all? Do you agree that they look slightly shifted on your plots? How long do these sims take, perhaps just do more simulations if these don't take very long? 
 - jovo said he wanted to see that the prob/rho values still looked right for a few other different values of rho and p.
 - have you tried any edge cases? Like what happens when rho is 1 or 0, or p is 1 or 0?

**Comment by SHAAAAN** [comment]
Hi Pedigo, I found a dimension mistake in the previous notebook and I revise it earlier today. Now I can see that the function works well. I also test the properties of the function when p ranging from 0.001 to 0.999 and rho ranging from 0.00 to 1.00. It seems still working well. (This jupyter notebook is the same link above.)

As for the Bear's codes, I have tested it for several pairs. But like we have discussed before, we have to define edge probability for both graphs. As the result, I haven't use his code so far.

**Comment by bdpedigo** [comment]
Hi @SHAAAAN 

> Hi @bdpedigo , I found a dimension mistake in the previous notebook and I revised it earlier today. Now we can see that the function works well. I also test the properties of the function when p ranging from 0.001 to 0.999 and rho ranging from 0.00 to 1.00. It seems still working well. (This jupyter notebook is the same link above.) In addition, I restore the recent four commits back to the previous version.

Thanks for updating the notebook. A few things: 
 - Somehow you are not calculating correlation of the sampled graphs correctly. I know this from your plots because correlation should be defined from [-1, 1] and your correlation looks like it 1) cannot be negative and 2) can be bigger than 0. I also don't see why you think the experiment with rho=0 worked, to me it look like the mean and expected are very different? 

> As for the Bear's codes, I have tested it for several pairs. But like we have discussed before, we have to define edge probability for both graphs. As the result, I haven't use his code so far.

I don't know what you mean about "define edge probability for both graphs." I can't quite remember the discussion that we had before. Does this mean that the two graphs can have different P matrices? If so, why can we not just set them the same and test against that?

**Comment by SHAAAAN** [comment]
Hi @bdpedigo , I have updated the above notebook and add the limitation of the relation between p and rho for er_corr and sbm_corr functions. And Bear's codes are also tested at the end of the above notebook. 
So far I have two "sklearn" errors, which I think is similar to Ali's. But other things seem to be correct.

File: graspy/simulations/simulations_corr.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    # check r
+    if not np.issubdtype(type(r), np.floating):
+        raise TypeError("r is not of type float.")
+    elif r < -1 or r > 1:
+        msg = "r must between -1 and 1."
+        raise ValueError(msg)
+
+    # check the relation between r and p
+    if p + r * (1 - p) < 0:
+        msg = "p + r * (1 - p) should be biiger than 0"
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: typo

File: graspy/simulations/simulations_corr.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        raise ValueError(msg)
+
+    if p * (1 - r) < 0:
+        msg = "p * (1 - r) should be biiger than 0"
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: typo

File: graspy/simulations/simulations_corr.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+    # check directed and loops
+    if type(directed) is not bool:
+        raise TypeError("directed is not of type bool.")
+    if type(loops) is not bool:
+        raise TypeError("loops is not of type bool.")
+
+    P = np.zeros((np.sum(n), np.sum(n)))
+    block_indices = np.insert(np.cumsum(np.array(n)), 0, 0)
+    p = np.mat(p)
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: why is this necessary?

File: tests/test_sims_corr.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+        # check rho
+        k1 = g1.copy()
+        k2 = g2.copy()
+        k1 = k1[np.where(~np.eye(k1.shape[0], dtype=bool))]
+        k2 = k2[np.where(~np.eye(k2.shape[0], dtype=bool))]
+        output_r = np.corrcoef(k1, k2)[0, 1]
+        self.assertTrue(np.isclose(self.r, output_r, atol=0.06))
+
+        # check the similarity of g1 and g2
```
Comment (bdpedigo) [inline tests/test_sims_corr.py]: is this redundant with checking the correlation?

Comment (bdpedigo) [reply tests/test_sims_corr.py]: i know you showed me at some point but i still don't really get what this is

**Comment by bdpedigo** [comment]
Hi Shan: 
corr.ipynb
- I think the paragraph at the beginning needs to be changed. I wouldn't talk about the implementation at all. I would basically just say that the correlated model means that we sample two graphs, but we can modify the probability of observing the same edge in both graphs by adjusting the correlation value. 
- for the titles of the first two heatmaps, just say Correlated ER Graph 1/2
- In the demo, I would only use the er_corr and sbm functions
- Can you remove the changes you made to erdos_reyni.ipynb and sbm.ipynb
- Have you tried to see if you can just do an import as `from graspy.simulations import er_corr`? That is what we want. 
Main code
- You have a lot of redundant code where you check inputs to er_corr and sbm_corr. Any common inputs (or parts of inputs that they share) you are welcome to make private functions to deal with them in both 
- Anywhere you have math or notationin the docstrings, you should get to render as math in sphinx. See examples in the rdpg code or RDPG estimator, it looks like `:math:` before some piece of text
- where is the most up to date proof notebook? Can you edit the PR description at the very top to include that, I would like to take another look and also have it linked to the PR itself. 

The code has come a really long way, nice work Shan!

File: tests/test_sims_corr.py (hunk 14 above)
Comment (SHAAAAN) [reply tests/test_sims_corr.py]: I can delete it if you think it is unnecessary. I just want to see how close the graph pair is.

File: graspy/simulations/simulations_corr.py (hunk 13 above)
Comment (SHAAAAN) [reply graspy/simulations/simulations_corr.py]: This is for the next few lines of making a generated P matrix to have the dimension of (n[0] + n[1]) * (n[0] + n[1]).

File: graspy/simulations/simulations_corr.py [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+            [0., 1., 0., 1., 1.],
+            [0., 0., 1., 0., 1.],
+            [0., 1., 1., 1., 0.]]))
+    """
+    # test input
+    # check P
+    if type(P) is not np.ndarray:
+        raise TypeError("P must be numpy.ndarray")
+    if len(P.shape) != 2:
+        raise ValueError("P must have dimension 2 (n_vertices, n_dimensions)")
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: (n_vertices, n_vertices)

File: graspy/simulations/simulations_corr.py [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    if P.shape[0] != P.shape[1]:
+        raise ValueError("P must be a square matrix")
+
+    # check R
+    if type(R) is not np.ndarray:
+        raise TypeError("R must be numpy.ndarray")
+    if len(R.shape) != 2:
+        raise ValueError("R must have dimension 2 (n_vertices, n_dimensions)")
```
Comment (bdpedigo) [inline graspy/simulations/simulations_corr.py]: same comment as above

File: graspy/simulations/simulations_corr.py (hunk 13 above)
Comment (bdpedigo) [reply graspy/simulations/simulations_corr.py]: I don't think you need to do `p = np.mat(p)` ?

Comment (SHAAAAN) [reply graspy/simulations/simulations_corr.py]: Sure. I can fix it.

**Comment by bdpedigo** [comment]
@SHAAAAN this looks pretty good. I want to play with the code a bit before merging but that's on me, not you. So for the purposes of class, no need to worry, I will try to merge it tomorrow but if I don't that's not your fault.

Nice work!
